tokio/src/sync/once_cell.rs

```diff
@@ -0,0 +1,313 @@
+        if self.initialized() {
+            // SAFETY: once the value is initialized, no mutable references are given out, so
+            // we can give out arbitrarily many immutable references
+            return unsafe { self.get_unchecked() };
+        } else {
+            // After acquire().await we have either acquired a permit while self.value
+            // is still uninitialized, or current thread is awoken after another thread
+            // has intialized the value and closed the semaphore, in which case self.initialized
+            // is true and we don't set the value here
+            match self.semaphore.acquire().await {
+                Ok(_permit) => {
+                    if !self.initialized() {
+                        // If `f` panics or `select!` is called, this `get_or_init` call
+                        // is aborted and the semaphore permit is dropped.
+                        let value = f.await;
+
+                        // SAFETY: There is only one permit on the semaphore, hence only one
+                        // mutable reference is created
+                        unsafe { self.set_value(value) };
+
+                        // SAFETY: once the value is initialized, no mutable references are given out, so
+                        // we can give out arbitrarily many immutable references
+                        return unsafe { self.get_unchecked() };
+                    } else {
+                        unreachable!("acquired semaphore after value was already initialized.");
+                    }
+                }
+                Err(_) => {
+                    if self.initialized() {
+                        // SAFETY: once the value is initialized, no mutable references are given out, so
+                        // we can give out arbitrarily many immutable references
+                        return unsafe { self.get_unchecked() };
+                    } else {
+                        unreachable!(
+                            "Semaphore closed, but the OnceCell has not been initialized."
+                        );
+                    }
+                }
+            }
+        }
+    }
+}
+
+// Since `get` gives us access to immutable references of the
+// OnceCell, OnceCell can only be Sync if T is Sync, otherwise
+// OnceCell would allow sharing references of !Sync values across
+// threads. We need T to be Send in order for OnceCell to by Sync
+// because we can use `set` on `&OnceCell<T>` to send
+// values (of type T) across threads.
+unsafe impl<T: Sync + Send> Sync for OnceCell<T> {}
+
+// Access to OnceCell's value is guarded by the semaphore permit
+// and atomic operations on `value_set`, so as long as T itself is Send
+// it's safe to send it to another thread
+unsafe impl<T: Send> Send for OnceCell<T> {}
+
+/// Error returned from the [`OnceCell::set`] method
+///
+/// [`OnceCell::set`]: crate::sync::OnceCell::set
+#[derive(Debug, PartialEq)]
+pub struct AlreadyInitializedError;
```

Comment:
This should have a private field to be consistent with the other errors in Tokio.

Reply:
What purpose does this private field have?

One problem I have is that for some reason we cannot instantiate tuple structs in the tests:

```
error[E0423]: expected function, tuple struct or tuple variant, found struct `NotInitializedError`
   --> tokio/tests/sync_once_cell.rs:88:28
    |
88  |     assert_eq!(uninit, Err(NotInitializedError(())));
    |                            ^^^^^^^^^^^^^^^^^^^^^^^ help: use struct literal syntax instead: `NotInitializedError { 0: val }`
    | 
   ::: /Users/bn/Documents/Programming/OpenSource/tokio-once-cell/tokio/src/sync/once_cell.rs:332:1
    |
332 | pub struct NotInitializedError(());
    | ----------------------------------- `NotInitializedError` defined here
```

Reply:
The private field has the purpose of preventing people from constructing the error. You can write the assert as
```Rust
assert!(uninit.is_err());
```